File: internal/game/facing.go
package game

import (
	"encoding/json"
	"math"
	"strconv"
	"strings"
)

// Facing yaw is a continuous Y-axis rotation in radians (Three.js model space).
//
//	0      north (−map Y / −world Z)
//	−π/2   east  (+map X)
//	 π     south (+map Y)
//	 π/2   west  (−map X)
const (
	FacingYawNorth   = 0.0
	FacingYawEast    = -math.Pi / 2
	FacingYawSouth   = math.Pi
	FacingYawWest    = math.Pi / 2
	FacingYawDefault = FacingYawSouth
)

// Legacy string constants — only for profile migration / old tests.
const (
	FacingLeft      = "left"
	FacingRight     = "right"
	FacingUp        = "up"
	FacingDown      = "down"
	FacingUpLeft    = "up-left"
	FacingUpRight   = "up-right"
	FacingDownLeft  = "down-left"
	FacingDownRight = "down-right"
)

// NormalizeYaw wraps to (−π, π].
func NormalizeYaw(yaw float64) float64 {
	if math.IsNaN(yaw) || math.IsInf(yaw, 0) {
		return FacingYawDefault
	}
	const twoPi = 2 * math.Pi
	yaw = math.Mod(yaw+math.Pi, twoPi)
	if yaw < 0 {
		yaw += twoPi
	}
	return yaw - math.Pi
}

// FacingYawFromDelta derives model yaw from map-space motion (+x right, +y down).
// Zero motion keeps current (default south).
func FacingYawFromDelta(dx, dy, current float64) float64 {
	if math.Hypot(dx, dy) < 1e-9 {
		return NormalizeYaw(current)
	}
	// Match Three.js facingYaw used by KayKit: modelYaw = −atan2(dx, −dy).
	return NormalizeYaw(-math.Atan2(dx, -dy))
}

// ResolveFacingYaw prefers a client-supplied yaw when hasClient is true.
func ResolveFacingYaw(dx, dy float64, clientYaw float64, hasClient bool, current float64) float64 {
	if hasClient {
		return NormalizeYaw(clientYaw)
	}
	return FacingYawFromDelta(dx, dy, current)
}

// FacingDir returns the unit map vector for a model yaw (+x right, +y down).
func FacingDir(yaw float64) (fx, fy float64) {
	yaw = NormalizeYaw(yaw)
	return -math.Sin(yaw), -math.Cos(yaw)
}

// FacingYawFromExit is the traveler's yaw when crossing a zone line.
func FacingYawFromExit(e MapExit, cols int) float64 {
	if cols <= 0 {
		cols = OverworldCols
	}
	if (e.MinC+e.MaxC)/2 >= cols/2 {
		return FacingYawEast
	}
	return FacingYawWest
}

// FacingYawFromLegacy maps old left/right/8-way strings to radians.
func FacingYawFromLegacy(s string) float64 {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case FacingUp, "north":
		return FacingYawNorth
	case FacingUpRight:
		return NormalizeYaw(-math.Pi / 4)
	case FacingRight, "east":
		return FacingYawEast
	case FacingDownRight:
		return NormalizeYaw(-3 * math.Pi / 4)
	case FacingDown, "south", "":
		return FacingYawSouth
	case FacingDownLeft:
		return NormalizeYaw(3 * math.Pi / 4)
	case FacingLeft, "west":
		return FacingYawWest
	case FacingUpLeft:
		return NormalizeYaw(math.Pi / 4)
	default:
		if v, err := strconv.ParseFloat(s, 64); err == nil {
			return NormalizeYaw(v)
		}
		return FacingYawDefault
	}
}

// ParseFacingYawJSON accepts a JSON number or legacy string.
func ParseFacingYawJSON(raw json.RawMessage, fallback float64) float64 {
	if len(raw) == 0 || string(raw) == "null" {
		return NormalizeYaw(fallback)
	}
	if raw[0] == '"' {
		var s string
		if err := json.Unmarshal(raw, &s); err != nil {
			return NormalizeYaw(fallback)
		}
		return FacingYawFromLegacy(s)
	}
	var v float64
	if err := json.Unmarshal(raw, &v); err != nil {
		return NormalizeYaw(fallback)
	}
	return NormalizeYaw(v)
}

// FacingYaw is a profile/persistable yaw with legacy string support.
type FacingYaw float64

func (f FacingYaw) Radians() float64 { return NormalizeYaw(float64(f)) }

func (f FacingYaw) MarshalJSON() ([]byte, error) {
	return json.Marshal(f.Radians())
}

func (f *FacingYaw) UnmarshalJSON(data []byte) error {
	*f = FacingYaw(ParseFacingYawJSON(data, FacingYawDefault))
	return nil
}

File: wails/frontend/src/three/facing.ts
/**
 * Continuous Y-axis facing (radians), matching Three.js model `rotation.y`.
 *
 *   0      north (−map Y)
 *  −π/2   east  (+map X)
 *   π     south (+map Y)
 *   π/2   west  (−map X)
 */
export type FacingYaw = number;

export const FACING_YAW_NORTH = 0;
export const FACING_YAW_EAST = -Math.PI / 2;
export const FACING_YAW_SOUTH = Math.PI;
export const FACING_YAW_WEST = Math.PI / 2;
export const FACING_YAW_DEFAULT = FACING_YAW_SOUTH;

/** @deprecated alias — prefer FacingYaw */
export type WorldFacing = FacingYaw;
export const WORLD_FACING_DEFAULT = FACING_YAW_DEFAULT;

export function normalizeYaw(yaw: number): number {
  if (!Number.isFinite(yaw)) return FACING_YAW_DEFAULT;
  const twoPi = Math.PI * 2;
  let y = ((yaw + Math.PI) % twoPi + twoPi) % twoPi;
  return y - Math.PI;
}

/** Parse wire facing: number radians, or legacy left/right/8-way string. */
export function parseFacingYaw(
  value: number | string | undefined | null,
  fallback: FacingYaw = FACING_YAW_DEFAULT,
): FacingYaw {
  if (typeof value === "number" && Number.isFinite(value)) return normalizeYaw(value);
  if (typeof value === "string") {
    const s = value.trim().toLowerCase();
    switch (s) {
      case "up":
      case "north":
        return FACING_YAW_NORTH;
      case "up-right":
        return normalizeYaw(-Math.PI / 4);
      case "right":
      case "east":
        return FACING_YAW_EAST;
      case "down-right":
        return normalizeYaw((-3 * Math.PI) / 4);
      case "down":
      case "south":
        return FACING_YAW_SOUTH;
      case "down-left":
        return normalizeYaw((3 * Math.PI) / 4);
      case "left":
      case "west":
        return FACING_YAW_WEST;
      case "up-left":
        return normalizeYaw(Math.PI / 4);
      default: {
        const n = Number(s);
        if (Number.isFinite(n)) return normalizeYaw(n);
      }
    }
  }
  return normalizeYaw(fallback);
}

/** @deprecated */
export function parseWorldFacing(
  value: number | string | undefined | null,
  fallback: FacingYaw = FACING_YAW_DEFAULT,
): FacingYaw {
  return parseFacingYaw(value, fallback);
}

/**
 * Model yaw from map-space motion (+x right, +y down).
 * Matches KayKit convention: modelYaw = −atan2(dx, −dy).
 */
export function facingFromAxes(dx: number, dy: number, current: FacingYaw): FacingYaw {
  if (dx === 0 && dy === 0) return normalizeYaw(current);
  return normalizeYaw(-Math.atan2(dx, -dy));
}

/** Identity — facing is already model yaw. */
export function facingYaw(facing: FacingYaw): number {
  return normalizeYaw(facing);
}

/** Unit vector on the map (+x right, +y down). */
export function facingDir(yaw: FacingYaw): { x: number; y: number } {
  const y = normalizeYaw(yaw);
  return { x: -Math.sin(y), y: -Math.cos(y) };
}

/** Pet follow offset behind the owner. */
export function followOffset(yaw: FacingYaw, dist: number): { x: number; y: number } {
  const d = facingDir(yaw);
  return { x: -d.x * dist, y: -d.y * dist };
}

/** Face along camera-forward (W / away-from-camera). Orbit yaw 0 → north. */
export function facingFromCameraYaw(cameraYaw: number): FacingYaw {
  return normalizeYaw(cameraYaw);
}
